### PR TITLE
functionaltest: Add one re-route processor test

### PR DESCRIPTION
**Pull request by ericywl**
## Changes

Based on https://github.com/elastic/apm-server/pull/15951
- Add `CreateRerouteProcessors` and `PerformManualRollovers` to ES client
- Add re-route processor test for 8.13.4 -> 8.16.0
- Parallelize functional tests

## How to test these changes

Run functional test on this branch.

Test run: https://github.com/elastic/apm-server/actions/runs/13675996351

### CODE REVIEW

**Comment by endorama** [comment]
Given that this PR is in draft, but is also doing a refactoring that may block other work (see https://github.com/elastic/apm-server/pull/16011 for example) can you please extract the refactoring in a separate PR so we can merge it faster?

**Comment by ericywl** [comment]
> Given that this PR is in draft, but is also doing a refactoring that may block other work (see https://github.com/elastic/apm-server/pull/16011 for example) can you please extract the refactoring in a separate PR so we can merge it faster?

I will move the refactoring to https://github.com/elastic/apm-server/pull/15951.

**Comment by ericywl** [comment]
Moved back to draft since it's failing.

File: functionaltests/9_0_test.go [hunk 1]
```diff
@@ -22,6 +22,7 @@ import (
 )
 
 func TestUpgrade_8_18_0_to_9_0_0(t *testing.T) {
+	t.Parallel()
```
Comment (endorama) [inline functionaltests/9_0_test.go:25]: let's do this in a separate PR

File: functionaltests/main_test.go [hunk 2]
```diff
@@ -62,14 +77,28 @@ func getDocsCountPerDS(t *testing.T, ctx context.Context, ecc *esclient.Client)
...
+	for _, v := range s {
+		m[v] = true
+	}
+	return m
+}
+
 // assertDocCount check if specified document count is equal to expected minus
 // documents count from a previous state.
-func assertDocCount(t *testing.T, docsCount, previous, expected esclient.APMDataStreamsDocCount) {
+func assertDocCount(t *testing.T, docsCount, previous, expected esclient.APMDataStreamsDocCount, skippedDataStreams []string) {
```
[the remark this answers is not in the document]
Comment (ericywl) [reply functionaltests/main_test.go:90]: In my case, it's so that we can assert if there are some unexpected data streams. It was helpful for the reroute test as it helped identify that I forgot to change the expected data streams to `rerouted` (i.e. the assertions weren't happening because it was trying to find `default`).

File: functionaltests/single_upgrade_test.go [hunk 3]
```diff
@@ -34,6 +34,37 @@ import (
...
+		ExpectedDocsCountPerIngestRun: expectedIngestForASingleRun(namespace),
+		SkippedDataStreams:            skippedIngest(namespace),
+	}
+}
+
+func newDefaultConfig() config {
+	return newDefaultConfigWithNamespace(defaultNamespace)
+}
+
+type additionalFn func(t *testing.T, ctx context.Context, cfg *config, deps dependencies) (continueTest bool)
```
[the remark this answers is not in the document]
Comment (ericywl) [reply functionaltests/single_upgrade_test.go:66]: In the case of reroute, I used this config to change the expected docs and skipped data streams midtest.

File: functionaltests/single_upgrade_test.go [hunk 4]
```diff
@@ -34,6 +34,37 @@ import (
 	"github.com/elastic/apm-server/functionaltests/internal/terraform"
 )
 
+type dependencies struct {
```
[the remark this answers is not in the document]
Comment (ericywl) [reply functionaltests/single_upgrade_test.go:37]: Sorry, not sure what you meant here. I think perhaps this `dependencies` was unnecessary, I could just pass ES client and Kibana client into the `additionalFn` separately. I added the struct to group them together, but if we don't have any other clients its probably unnecessary.

File: functionaltests/single_upgrade_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (endorama) [reply functionaltests/single_upgrade_test.go:66]: I see now, sorry for having miss that earlier. It looks a bit complex from what it achieves, what if instead we add a `namespace` field to the test struct? We can then initialize the expected values in the test, with using the empty value to default to `default`, maintaining the current behavior.

> I'm not sure if there are future tests that would need changing the expected docs etc. mid-test.

That's a valid concern but I would address it once we have a valid use case. So we can keep it simpler until needed.

File: functionaltests/single_upgrade_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (endorama) [reply functionaltests/single_upgrade_test.go:37]: You're right, we cannot do what I propose without adding a factory function for the test struct.
As you suggest I would remove the `dependencies` struct. I don't foresee more clients to be needed here (_maybe_ one more, but no more than that).

File: functionaltests/main_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (endorama) [reply functionaltests/main_test.go:90]: That make perfect sense, good call.

File: functionaltests/single_upgrade_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (ericywl) [reply functionaltests/single_upgrade_test.go]: Makes sense, that was my original approach 😅. Will make the change back.

File: functionaltests/single_upgrade_test.go [hunk 5]
```diff
@@ -34,6 +34,8 @@ import (
 	"github.com/elastic/apm-server/functionaltests/internal/terraform"
 )
 
+type additionalFunc func(t *testing.T, ctx context.Context, esc *esclient.Client, kbc *kbclient.Client) (continueTest bool)
```
Comment (endorama) [inline functionaltests/single_upgrade_test.go:37]: What do you think returning an error here and the asserting `require.NoError` on line 92? This would allow to pass back more contextual information on why it failed.

**Review by endorama — DISMISSED** [review]
A minor non blocking suggestion but this looks ready for merge! Feel free to address it in a follow up.

File: functionaltests/single_upgrade_test.go (hunk 5 above)
Comment (ericywl) [reply functionaltests/single_upgrade_test.go:37]: Sounds good, I will make the change here so that there's no conflicts in the follow ups.

**Review by endorama — APPROVED** [review]
The last CI run failed due to the 9.0.0 build issues. But the `TestUpgrade_8_13_4_to_8_16_0_Reroute` [passed](https://github.com/elastic/apm-server/actions/runs/13699469406/job/38309410983#step:7:160), so it's good to merge for me.
